src/external.jl

```diff
@@ -0,0 +1,152 @@
+function fix_external_callback end
+
+mutable struct FixExternal
+    const lmp::LMP
+    const name::String
+    const callback
+    timestep::Int
+    nlocal::Int
+    ids::Union{Nothing, Vector{Cint}}
+    x::Union{Nothing, Matrix{Float64}}
+    f::Union{Nothing, Matrix{Float64}}
+
+    function FixExternal(callback, lmp::LMP, name::String, group::String, ncall::Int, napply::Int)
+        command(lmp, "fix $name $group external pf/callback $ncall $napply")
+        this = new(lmp, name, callback, 0, 0, nothing, nothing, nothing)
+        lmp.external_fixes[name] = this # preserves pair globally
+
+        ctx = Base.pointer_from_objref(this)
+        callback = @cfunction(fix_external_callback, Cvoid, (Ptr{Cvoid}, Int64, Cint, Ptr{Cint}, Ptr{Ptr{Float64}}, Ptr{Ptr{Float64}}))
+        API.lammps_set_fix_external_callback(lmp, name, callback, ctx)
+
+        # Ensure function is compiled before timestep 0
+        if !precompile(this.callback, (FixExternal,))
+            @warn "Failed to precompile the callback" this.callback
+        end
+        return this
+    end
+end
+
+function fix_external_callback(ctx::Ptr{Cvoid}, timestep::Int64, nlocal::Cint, ids::Ptr{Cint}, x::Ptr{Ptr{Float64}}, fexternal::Ptr{Ptr{Float64}})
+    fix = Base.unsafe_pointer_to_objref(ctx)::FixExternal
+    fix.timestep = timestep
+    fix.nlocal = nlocal
+
+    nall = extract_setting(fix.lmp, "nall")
+    fix.x = _extract(x, (Int32(3), nall))
+    fix.f = _extract(fexternal, (Int32(3), nall))
+    fix.ids = _extract(ids, nall)
+
+    # necessary dynamic
+    fix.callback(fix)
+
+    fix.timestep = 0
+    fix.nlocal = 0
+    fix.x = nothing
+    fix.f = nothing
+    fix.ids = nothing
+
+    return nothing
+end
+
+function set_virial_peratom(fix::FixExternal, virial::AbstractVector{SVector{6, Float64}}; set_global=false)
+    @assert length(virial) >= fix.nlocal
+
+    @no_escape begin
+        ptrs = @alloc(Ptr{Ptr{Float64}}, fix.nlocal)
+        for i in eachindex(ptrs)
+            @inbounds ptrs[i] = pointer(virial, i)
+        end
+        API.lammps_fix_external_set_virial_peratom(fix.lmp, fix.name, ptrs)
+    end
+
+    if set_global
+        virial_global = MVector(sum(virial))
+        comm = API.lammps_get_mpi_comm(fix.lmp)
```

Comment:
FFS...

I think on OpenMPI & co you will need `MPI.API.MPI_Comm_f2c(comm)` since we use the C